File: auntieos-admin/web/composeApp/src/commonMain/kotlin/com/tribetails/auntieos/web/data/ChecklistBank.kt
package com.tribetails.auntieos.web.data

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive

/**
 * Run-4 #7b: one item in the shared bank of common KinTale checklist tasks. The bank
 * backs the editor's "Add from bank" picker; admins grow it with "Save to bank".
 */
@Serializable
data class ChecklistBankItem(
    val id: String = "",
    val text: String = "",
    val scope: String = "PER_PET",          // "PER_PET" | "PER_VISIT"
)

private val checklistBankJson = Json { ignoreUnknownKeys = true; isLenient = true }

/** Pure decode of the listChecklistBank body `{items:[{id,text,scope}]}`. */
internal fun decodeChecklistBank(dataJson: String): List<ChecklistBankItem> =
    (checklistBankJson.parseToJsonElement(dataJson).jsonObject["items"] as? JsonArray)
        .orEmpty()
        .mapNotNull { el ->
            val o = el as? JsonObject ?: return@mapNotNull null
            ChecklistBankItem(
                id = o["id"]?.jsonPrimitive?.contentOrNull.orEmpty(),
                text = o["text"]?.jsonPrimitive?.contentOrNull.orEmpty(),
                scope = o["scope"]?.jsonPrimitive?.contentOrNull?.takeIf { it == "PER_VISIT" } ?: "PER_PET",
            )
        }
        .filter { it.text.isNotBlank() }

private fun norm(s: String): String = s.trim().lowercase()

/** True when [items] already contains an item with the given text (case/space-insensitive). */
fun checklistHasText(items: List<ChecklistItem>, text: String): Boolean {
    val key = norm(text)
    return items.any { norm(it.text) == key }
}

/**
 * Bank items not already in the template's checklist, so the picker never offers a
 * duplicate of an item the admin has already added.
 */
fun bankItemsNotInChecklist(bank: List<ChecklistBankItem>, items: List<ChecklistItem>): List<ChecklistBankItem> =
    bank.filterNot { checklistHasText(items, it.text) }

/** Build a template [ChecklistItem] from a picked bank item. */
fun checklistItemFromBank(bankItem: ChecklistBankItem, key: String, order: Int): ChecklistItem =
    ChecklistItem(
        key = key,
        text = bankItem.text,
        scope = if (bankItem.scope == "PER_VISIT") "PER_VISIT" else "PER_PET",
        order = order,
    )

File: auntieos-admin/src/lib/checklistBank.ts
import { makeChecklistItem, type ChecklistItem } from './kinTale/model';

/**
 * The shared bank of common KinTale checklist items: the pure half, ported
 * case-for-case from the archive's `data/ChecklistBank.kt` (and matching
 * android's `ui/kintales/ChecklistBankHelpers.kt`, so the same item offered on
 * either platform lands identically).
 *
 * The I/O half is `api/checklistBank.ts` (the `listChecklistBank` /
 * `saveChecklistBankItem` admin callables). Kept apart so the picker's
 * de-duplication and scope rules are testable without a callable in sight.
 */

export type BankScope = 'PER_PET' | 'PER_VISIT';

/** One saved bank row. `scope` is narrowed on decode, so it is never a third value. */
export interface ChecklistBankItem {
  id: string;
  text: string;
  scope: string;
}

function norm(s: string): string {
  return (s ?? '').trim().toLowerCase();
}

/**
 * Narrows a free-text scope to the two the bank models. Anything unrecognized
 * becomes PER_PET, matching the Kotlin decoder's
 * `takeIf { it == "PER_VISIT" } ?: "PER_PET"` and the callable's zod default.
 */
export function normalizeBankScope(scope: string): BankScope {
  return norm(scope) === 'per_visit' ? 'PER_VISIT' : 'PER_PET';
}

/** True when [items] already carries this text, ignoring case and surrounding space. */
export function checklistHasText(items: readonly ChecklistItem[], text: string): boolean {
  const key = norm(text);
  return items.some((i) => norm(i.text) === key);
}

/**
 * Bank rows not already present in the template's checklist, so the quick-add
 * never offers a duplicate of something the operator has already added.
 */
export function bankItemsNotInChecklist(
  bank: readonly ChecklistBankItem[],
  items: readonly ChecklistItem[],
): ChecklistBankItem[] {
  return bank.filter((b) => !checklistHasText(items, b.text));
}

/**
 * What one checklist SECTION should offer: the not-yet-added bank rows saved for
 * that scope. Scope is compared after normalization, so a legacy lowercase
 * `per_visit` row still lands in the per-visit section rather than silently
 * defaulting into per-pet.
 */
export function bankItemsForScope(
  bank: readonly ChecklistBankItem[],
  items: readonly ChecklistItem[],
  scope: string,
): ChecklistBankItem[] {
  const want = normalizeBankScope(scope);
  return bankItemsNotInChecklist(bank, items).filter((b) => normalizeBankScope(b.scope) === want);
}

/**
 * Build a template checklist row from a picked bank item. Everything except
 * text/scope/key/order comes from the model default: a bank item carries no
 * visibility conditions, so the new row starts unconditionally visible.
 */
export function checklistItemFromBank(
  bankItem: ChecklistBankItem,
  key: string,
  order: number,
): ChecklistItem {
  return makeChecklistItem({
    key,
    text: bankItem.text,
    scope: normalizeBankScope(bankItem.scope),
    order,
  });
}
